x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/action/ShardChangesAction.java

```diff
@@ -308,14 +323,15 @@ protected Response shardOperation(Request request, ShardId shardId) throws IOExc
                     request.getMaxBatchSize());
             // must capture after after snapshotting operations to ensure this MUS is at least the highest MUS of any of these operations.
             final long maxSeqNoOfUpdatesOrDeletes = indexShard.getMaxSeqNoOfUpdatesOrDeletes();
-            return getResponse(mappingVersion, seqNoStats, maxSeqNoOfUpdatesOrDeletes, operations);
+            return getResponse(mappingVersion, seqNoStats, maxSeqNoOfUpdatesOrDeletes, operations, request.relativeStartNanos);
         }
 
         @Override
         protected void asyncShardOperation(
                 final Request request,
                 final ShardId shardId,
                 final ActionListener<Response> listener) throws IOException {
+            request.relativeStartNanos = System.nanoTime();
```

Comment:
I think this is the wrong place to start the clock. The goal with this metric is to capture the non-network time on the leader so we want to start the clock as soon as possible. Right now these requests are handled on the network thread before we go async, so there is not much in terms of queuing time, but it would be a bug if somehow this ever changed and handling of this request was not on the network thread before going async, this would miss queuing time in that other thread pool (we miss internal Netty queuing, but I don't think there's anything obvious that we can do about that). This is also something to think about in the context of the NIO work. I also don't like that we are mutating the request, it's doing to make it harder to have immutable request objects.

I think we should start the clock the moment the request is de-serialized.